tools/halide_image_io.h

```diff
@@ -1048,6 +1048,369 @@ bool save_ppm(ImageType &im, const std::string &filename) {
+constexpr bool host_is_big_endian = false;
+#endif
+
+constexpr char little_endian_char = '<';
+constexpr char big_endian_char = '>';
+constexpr char no_endian_char = '|';
+constexpr char host_endian_char = (host_is_big_endian ? big_endian_char : little_endian_char);
+
+struct npy_dtype_info_t {
+    char byte_order;
+    char kind;
+    size_t item_size;
+
+    std::string descr() const {
+        return std::string(1, byte_order) + std::string(1, kind) + std::to_string(item_size);
+    }
+};
+
+inline static const std::array<std::pair<halide_type_t, npy_dtype_info_t>, 10> npy_dtypes = {{
+    // TODO: float16
+    {halide_type_of<float>(), {host_endian_char, 'f', sizeof(float)}},
+    {halide_type_of<double>(), {host_endian_char, 'f', sizeof(double)}},
+    {halide_type_of<int8_t>(), {no_endian_char, 'i', sizeof(int8_t)}},
+    {halide_type_of<int16_t>(), {host_endian_char, 'i', sizeof(int16_t)}},
+    {halide_type_of<int32_t>(), {host_endian_char, 'i', sizeof(int32_t)}},
+    {halide_type_of<int64_t>(), {host_endian_char, 'i', sizeof(int64_t)}},
+    {halide_type_of<uint8_t>(), {no_endian_char, 'u', sizeof(uint8_t)}},
+    {halide_type_of<uint16_t>(), {host_endian_char, 'u', sizeof(uint16_t)}},
+    {halide_type_of<uint32_t>(), {host_endian_char, 'u', sizeof(uint32_t)}},
+    {halide_type_of<uint64_t>(), {host_endian_char, 'u', sizeof(uint64_t)}},
+}};
+
+inline static const std::array<char, 6> npy_magic_string = {'\x93', 'N', 'U', 'M', 'P', 'Y'};
+inline static const std::array<char, 2> npy_v1_bytes = {'\x01', '\x00'};
+
+inline std::string trim_whitespace(const std::string &s) {
+    const size_t first = s.find_first_not_of(" \t\n");
+    if (first == std::string::npos) {
+        return "";
+    }
+    const size_t last = s.find_last_not_of(" \t\n");
+    return s.substr(first, (last - first + 1));
+}
+
+inline bool parse_python_dict(std::string dict_str, std::map<std::string, std::string> &m) {
+    dict_str = trim_whitespace(dict_str);
+
+    if (dict_str.front() != '{' || dict_str.back() != '}') {
+        return false;  // not a python dict
+    }
+    dict_str = dict_str.substr(1, dict_str.length() - 2);
+
+    std::vector<std::pair<size_t, std::string>> pos;
+
+    std::vector<std::pair<size_t, std::string>> positions;
+
+    constexpr int kKeyCount = 3;
+    std::string keys[kKeyCount] = {"descr", "fortran_order", "shape"};
+    for (const auto &k : keys) {
+        size_t pos = dict_str.find("'" + k + "'");
+        if (pos == std::string::npos) {
+            return false;  // missing a required key
+        }
+        positions.emplace_back(pos, k);
+    }
+
+    // The keys aren't guaranteed to be sorted, but our loop below
+    // needs them to be low-to-high
+    std::sort(positions.begin(), positions.end());
+
+    // Here we are going to slurp everything from the start of a key
+    // to just before the start of the next key (or end of input),
+    // then split on : with everything to the right as the presumed value.
```

Comment:
So if there's an unusual key in there, this will be wrong. Does the spec guarantee that only those three keys exist for the versions we support?

Reply:
It's not clear; the spec says `The dictionary contains three keys` but doesn't say it contains *only* these.

Reply:
(BTW, I have verified that a `.npy` file produced by NumPy in Python on my Mac loads fine here... sample size = 1 :-)